### PR TITLE
Add S3 Storage Engine

### PR DESCRIPTION
**Pull request by ehsanmasdar**

### CODE REVIEW

File: server/storage.ts [hunk 1]
```diff
@@ -59,11 +61,68 @@ class DiskStorageEngine implements IStorageEngine {
...
+class S3StorageEngine implements IStorageEngine {
+	public readonly uploadRoot: string;
+	private readonly options: IS3Options;
+
+	constructor(options: IS3Options) {
+		// Values copied via spread operator instead of being passed by reference
+		this.options = {
+			...options
+		};
+		this.uploadRoot = this.options.uploadDirectory;
```
Comment (petschekr) [inline server/storage.ts]: Is there anything you have to do here to ensure that the bucket exists / create it?

Comment (ehsanmasdar) [inline server/storage.ts]: @petschekr No. Buckets are pretty static and the account in question should not even have permission to modify bucket-level settings in S3.

Comment (petschekr) [inline server/storage.ts]: Got it

Comment (ehsanmasdar) [inline server/storage.ts]: Whoever is specifying the bucket here should already have created it with their admin-level privileges, not expect the app to create the bucket for them

**Pull request merged**